src/tox/session/commands/run/parallel.py

```diff
@@ -82,25 +84,39 @@ def run_in_thread(tox_env, os_env):
                     reporter.quiet(message)
 
         threads = []
+        processes = []
         todo_keys = set(venv_dict.keys())
         todo = OrderedDict((n, todo_keys & set(v.envconfig.depends)) for n, v in venv_dict.items())
         done = set()
-        while todo:
-            for name, depends in list(todo.items()):
-                if depends - done:
-                    # skip if has unfinished dependencies
+        try:
+            while todo:
+                for name, depends in list(todo.items()):
+                    if depends - done:
+                        # skip if has unfinished dependencies
+                        continue
+                    del todo[name]
+                    venv = venv_dict[name]
+                    semaphore.acquire(blocking=True)
+                    spinner.add(name)
+                    thread = Thread(
+                        target=run_in_thread, args=(venv, os.environ.copy(), processes)
+                    )
+                    thread.start()
+                    threads.append(thread)
+                if todo:
+                    # wait until someone finishes and retry queuing jobs
+                    finished.wait()
+                    finished.clear()
+            for thread in threads:
+                thread.join()
+        except KeyboardInterrupt:
+            while True:
+                # do not allow to interrupt until children interrupt
+                try:
+                    for proc in list(processes):
+                        proc.send_signal(signal.CTRL_C_EVENT if sys.platform == "win32" else signal.SIGINT)
+                    for thread in threads:
+                        thread.join()
+                except KeyboardInterrupt:
                     continue
```

Comment:
I think if you run the same bit I pasted in https://github.com/tox-dev/tox/issues/1172#issuecomment-471071464 this'll now be unkillable with ^C

(my reproduction was for non-parallel mode as well -- not sure if that gets covered here or not?)

Reply:
Indeed, but if we don't do this like this then we cannot guarantee child elements will be able to clean-up, is that what we want?

Reply:
Indeed, was thinking of allowing children to clean-up, but on second thought let's be forceful and terminate instead.

Reply:
The usual pattern is INT -> TERM -> KILL (with some grace period in between)

Reply:
Yeah, just done that. 👍

Reply:
I see a proc.terminate(), but I don't see a proc.kill() anywhere?

Reply:
The terminate will kill not?

Reply:
@asottile suggested

> The usual pattern is INT -> TERM -> KILL (with some grace period in between)

You currently have the first two steps implemented, but your "yeah, just done that" comment gave me the impression you intended to implement all three?  

The reason for SIGKILL is that (buggy) programs can capture and ignore SIGTERM, but they cannot ignore SIGKILL.

Reply:
(TBH I personally don't think there's need for both SIGINT and SIGTERM; either would do, when followed by SIGKILL.  But I'm not a Unix guru.)

(Also, I wonder if it's enough to kill direct children, as they might have child processes of their own.  Setting up a new process group and sending signals to it might be better?  Again, not a Unix guru.)

Reply:
at least when the child process has a high probability of being a python process sending `INT` first is polite -- python (like most languages) doesn't attach a signal handler for `TERM` by default (finally blocks won't run, etc.) but they will with `INT` (as it gets converted to a `KeyboardInterrupt` exception)

Reply:
it's safe to add kill then for the general case, will amend 👍